### ISSUE TITLE
Monitors swapped on unlock

### ISSUE THREAD

**Issue by olantwin** [paperwm/PaperWM#565]
Hi, thanks for the great work! PaperWM is so central to my workflow these days, that I'm happy that development has picked up again!

Please let me know if I can do anything to help debug this.

**Describe the bug**

As of the most recent update, every unlock seems to swap the primary and secondary monitor. This used to not happen in 44.4 (not entirely sure when this regression started as I rarely completely log out of gnome shell, but update my clone of PaperWM every release).

**To Reproduce**

1. Have e.g. workspace 1 on primary monitor, workspace 3 on secondary
2. Lock gnome-shell
3. On unlock, workspace 1 will be on the secondary monitor and workspace 3 on primary

**Expected behavior**

Workplaces stay on the same monitor as before locking

**System information:**

```
Distribution: Arch Linux
GNOME Shell 44.3
Display server: Wayland
PaperWM branch/tag: v44.5.2
PaperWM commit: d89bb11cc8e63e775ee414ffc81e1c4caff9e127
Enabled extensions:
- caffeine@patapon.info
- paperwm@paperwm.github.com
- GoogleEarthWallpaper@neffo.github.com
- trayIconsReloaded@selfmade.pl
- appindicatorsupport@rgcjonas.gmail.com
- gsconnect@andyholmes.github.io
- switcher@landau.fi
```

**Comment by jtaala** [comment]
Hey @olantwin,

Ah, yep, so there's been some changes related to this: https://github.com/paperwm/PaperWM/pull/560#issuecomment-1646443791

I'll give reproducing this a go tonight.

Cheers.

**Comment by jtaala** [comment]
Ah, yep can reproduce - that comment above solved the "windows to right" workspaces" issue - but looks like there is "workspace to right/ monitor" issue.

Could reproduce, after playing around with workspaces (e.g. switching around workspaces on a monitor).  Looks like is currentyl remembering the original workspaces to monitor linking (but in PaperWM you can switch that around easily).

Will look at a fix and push it tonight.

Note, this is caused by a strict requriement by extensions.gnome.org (ego) re sessions modes (previously paperwm kept alive during lockscreens... can't do that if we want to add paperwm ego, so working though issues there).  Thanks for reporting!

P.S. feel free to jump back on release `v44.4` until we can get that fixed.

Jay.

**Comment by olantwin** [comment]
Hi Jay,

Thanks for the quick reply and reproducing it so quickly.

> Note, this is caused by a strict requriement by extensions.gnome.org (ego) re sessions modes (previously paperwm kept alive during lockscreens... can't do that if we want to add paperwm ego, so working though issues there). Thanks for reporting!

This makes sense. Are EGO extensions allowed to keep state somewhere on disk? Might not be the most elegant workaround, but would be a fallback if no better fix can be found.

Please let me know if there is anything I can do to help or to test possible fixes!

Cheers,
Oliver

**Comment by jtaala** [comment]
Hey @olantwin - I've added a PR which should fix this and a few other issues.  Are you able to test this a bit?

```bash
git fetch --all
git checkout fix-multimonitor-correct-workspace-on-unlock
./install.sh
```

and then logout/login.  Give locking/unlocking a bunch of goes - it should maintain (restore) the pre-locked workspace/monitor layouts after unlock.

**Comment by olantwin** [comment]
That PR seems to fix the problem. I'll test a bit more throughout the day, but so far I have not found any side effects of the fix either.

Thanks a lot @jtaala!

**Comment by jtaala** [comment]
Hey @olantwin - so I'm ready to merge this fix in.  I've actually fixed quite a few issues re multimonitors.
Also, I (re)fixed the window positioning code so it now works in multimonitors (actually, I accidentally removed saving window positions when I implemented the monitor fixes).

**Comment by jtaala** [comment]
Fixed by #567.
